exchange/exchange.go

```diff
@@ -189,52 +167,79 @@ func (e *exchange) HoldAuction(ctx context.Context, bidRequest *openrtb.BidReque
 				errs = append(errs, dealErrs...)
 			}
 
-			if debugLog != nil && debugLog.Enabled {
```

Comment:
There's a lot going on here. Can you please describe why you moves the debug log code?

Reply:
In the current version of `HoldAuction` this `debugLog != nil && debugLog.Enabled` condition gets evaluated regardless of `anyBidsReturned` being `true` or `false` (lines 192 and 218 of current code). Therefore, it made sense to evaluate this expression regardless of `anyBidsReturned`. This PR is changing the order and calling `bidResponseExt = e.makeExtBidResponse(adapterBids...)` regardless of `debugLog != nil && debugLog.Enabled` because, in the end, even if `debugLog != nil && debugLog.Enabled` was `false`, `bidResponseExt = e.makeExtBidResponse(adapterBids...)` gets called inside `buildBidResponse(..)` anyway (line 503 current code).

I put toghether a truth table in my notes and, even if the **Before** and **After** tables are not identical, I believe the **After** table still makes sense:

**Before**
```
   |                 |                 | debugLog != nil && | if rawUUID, err :=  ||                        | bidRespExtBytes, err := json.Marshal(bidResponseExt) 
   | anyBidsReturned | targData != nil |  debugLog.Enabled  |       uuid.NewV4(); || e.makeExtBidResponse() | (Not called if debugLog != nil && debugLog.Enabled is false)
---|-----------------|-----------------|--------------------|---------------------||------------------------|-------------------------------------------------------------
  1|        F        |  not called     |   line 218    F    |    not called       ||     line 509           |  not called 
  2|        F        |  not called     |   line 218    F    |    not called       ||     line 509           |  not called 
  3|        F        |  not called     |   line 218    T    |    line 219 F       ||     line 509           |  not called (rawUUID, err := uuid.NewV4(); is false)
  4|        F        |  not called     |   line 218    T    |    line 219 T       ||     line 222           |  line 223
  5|        F        |  not called     |   line 218    F    |    not called       ||     line 509           |  not called 
  6|        F        |  not called     |   line 218    F    |    not called       ||     line 509           |  not called 
  7|        F        |  not called     |   line 218    T    |    line 219 F       ||     line 509           |  not called (rawUUID, err := uuid.NewV4(); is false)
  8|        F        |  not called     |   line 218    T    |    line 219 T       ||     line 222           |  line 223
  9|        T        |  line 184   F   |    not called      |    not called       ||     line 509           |  not called  
 10|        T        |  line 184   F   |    not called      |    not called       ||     line 509           |  not called  
 11|        T        |  line 184   F   |    not called      |    not called       ||     line 509           |  not called
 12|        T        |  line 184   F   |    not called      |    not called       ||     line 509           |  not called
 13|        T        |  line 184   T   |   line 192    F    |    not called       ||     line 509           |  not called  
 14|        T        |  line 184   T   |   line 192    F    |    not called       ||     line 509           |  not called  
 15|        T        |  line 184   T   |   line 192    T    |    not called       ||     line 193           |  line 194 
 16|        T        |  line 184   T   |   line 192    T    |    not called       ||     line 193           |  line 194 
```

**After**
```
   |                 | targData != nil | debugLog != nil && | if rawUUID, err :=  ||                        | bidRespExtBytes, err := json.Marshal(bidResponseExt)
   | anyBidsReturned |  IRRELEVANT     |  debugLog.Enabled  |       uuid.NewV4(); || e.makeExtBidResponse() | (Not called if debugLog != nil && debugLog.Enabled is false)
---|-----------------|-----------------|--------------------|---------------------||------------------------|-------------------------------------------------------------
  1|        F        |  not called     |   line 180    F    |    not called       ||     line 179           |   not called 
  2|        F        |  not called     |   line 180    F    |    not called       ||     line 179           |   not called 
  3|        F        |  not called     |   line 180    T    |    line 188 F       ||     line 179           | + line 181 (we simply don't debugLog.CacheKey = rawUUID.String())
  4|        F        |  not called     |   line 180    T    |    line 188 T       ||     line 179           |   line 181
  5|        F        |  not called     |   line 180    F    |    not called       ||     line 179           |   not called 
  6|        F        |  not called     |   line 180    F    |    not called       ||     line 179           |   not called 
  7|        F        |  not called     |   line 180    T    |    line 188 F       ||     line 179           | + line 181 (we simply don't debugLog.CacheKey = rawUUID.String())
  8|        F        |  not called     |   line 180    T    |    line 188 T       ||     line 179           |   line 181
  9|        T        |  line 162   F   | + line 180    F    |    not called       ||     line 179           |   not called  
 10|        T        |  line 162   F   | + line 180    F    |    not called       ||     line 179           |   not called  
 11|        T        |  line 162   F   | + line 180    T    |    not called       ||     line 179           |   line 181
 12|        T        |  line 162   F   | + line 180    T    |    not called       ||     line 179           |   line 181
 13|        T        |  line 162   T   |   line 180    F    |    not called       ||     line 179           |   not called  
 14|        T        |  line 162   T   |   line 180    F    |    not called       ||     line 179           |   not called  
 15|        T        |  line 162   T   |   line 180    T    |    not called       ||     line 179           |   line 181
 16|        T        |  line 162   T   |   line 180    T    |    not called       ||     line 179           |   line 181
```

I don't believe anything gets broken and the fact that all tests are passing, suggests it. The discrepancies in scenarios 9 through 12 of the `debugLog != nil && debugLog.Enabled` column seem to make sense IMHO, as well as rows 3 and 7 of the `bidRespExtBytes, err := json.Marshal(bidResponseExt)` column.

But please let me know your thoughts on this. Do you think this logic should be rearranged to make the tables match?

Reply:
Gotcha. Thank you for the background info.